#ml GCC OpenMPI R Anaconda3/2022.05; R

args<-as.numeric(commandArgs(TRUE))

library(optparse)
library(glmnet)
library(data.table)
library(foreach) 
library(doParallel)
library(tidyr)
library(dbplyr)

#---defaul settings---
parallel = TRUE
n_genes_per_subjob = 500
tissue = 'geuvadis'
n_tissue = 421

#---set up sub jobs---
alpha = '1'

#---for test---
# alpha = '0.5'
# pvalue_cutoff = 'pval_0.05'
# subjob_id = 1

#prscs path
prscs = './tools/PRScs/PRScs.py'
#ld reference for prscs
ld_ref = c("./tools/PRS-CS/ref/ldblk_1kg_eur")

#main dir
main_dir = paste0("./summary_stat_prediction/geuvadis/prscs_weights/")

#ref in plink format
plink_file_name = "./projects/cross_tissue/rep/data/geno/geuvadis421"


#mkdir
if(!dir.exists(main_dir)){
  dir.create(main_dir)
}
if(!dir.exists(paste0(main_dir,'/gene/'))){
  dir.create(paste0(main_dir,'/gene/'))
} 
if(!dir.exists(paste0(main_dir,'/weights/'))){
  dir.create(paste0(main_dir,'/weights/'))
}  
if(!dir.exists(paste0(main_dir,'/cov/'))){
  dir.create(paste0(main_dir,'/cov/'))
}
if(!dir.exists(paste0(main_dir,'/tmp/'))){
  dir.create(paste0(main_dir,'/tmp/'))
}


#load eQTL data
print('INFO loading eQTL data ...')
eqtl = fread("./summary_stat_prediction/geuvadis/eqtl/geuvadis_eQTL_with_info.txt") %>% as.data.frame()

#gene list
gene_list <- eqtl$Gene[!duplicated(eqtl$Gene)]

#main fun
each_gene <- function(gene_id){
  
  #select eqtl for the target gene
  tmp_eqtl <- eqtl[eqtl$Gene == gene_id,]
  
  if(nrow(tmp_eqtl) >= 2){
    
    #generate genotype file
    write.table(tmp_eqtl$rsid, row.names=F, col.names=F, quote=F, 
                file=paste0(main_dir,"/tmp/",gene_id,"_rs_id.txt"))
    cmd1<-paste0('plink --bfile ',plink_file_name, " --extract ",main_dir,"/tmp/",gene_id,"_rs_id.txt" , " --recodeA --out ",main_dir,'/gene/',gene_id)
    system(cmd1,ignore.stdout=T,ignore.stderr=T,wait=T)
    
    if(file.exists(paste0(main_dir,'/gene/',gene_id,".raw"))){
      dosage_raw <- fread(paste0(main_dir, "/gene/",gene_id,".raw")) %>% as.data.frame()
      
      if(ncol(dosage_raw)>8){
        
        #prepare prscs input file
        dosage <- dosage_raw[,-c(1,3:6)]
        rownames(dosage) <- dosage$IID
        dosage <- dosage[,-1]
        
        #keep the counted alleles in .raw data from plink reference data
        counted_allele = as.character(sapply(colnames(dosage), function(x) strsplit(x,"\\_")[[1]][2])) ###
        
        colnames(dosage) <- as.character(sapply(colnames(dosage), function(x) strsplit(x,"\\_")[[1]][1]))
        
        #take overlapped snps for dosage and tmp_eqtl
        tmp_eqtl <- tmp_eqtl[tmp_eqtl$rsid %in% colnames(dosage),]
        tmp_summ <- tmp_eqtl[,c('rsid','eff_allele','ref_allele','beta','pval')]
        colnames(tmp_summ) <- c('SNP','A1','A2','BETA','P')
        
        #generate prscs input files
        tmp_bim <- data.frame(tmp_eqtl$chr,tmp_eqtl$rsid,0,tmp_eqtl$pos,tmp_eqtl$eff_allele,tmp_eqtl$ref_allele)
        data.table::fwrite(tmp_bim, sep = "\t", row.names = F, col.names = F, quote = F,
                           file = paste0(main_dir,"/gene/",gene_id,".bim"))
        data.table::fwrite(tmp_summ, sep = "\t", row.names = F, col.names = T, quote = F,
                           file = paste0(main_dir,"/gene/",gene_id,".eqtl_summary"))
        
        #run prscs
        gene_plink <- paste0(main_dir,'/gene/',gene_id)
        prs_eqtl   <- paste0(main_dir,'/gene/',gene_id,".eqtl_summary")
        sample_saize <- n_tissue
        chr <- tmp_eqtl[1,3]
        
        cmd2 <- paste0("python ",prscs, " --ref_dir=",ld_ref, " --bim_prefix=",gene_plink, 
                       " --sst_file=", prs_eqtl," --n_gwas=",sample_saize , " --chr=",chr,
                       " --out_dir=",main_dir,"/weights/", gene_id, " --n_iter=4000 --a=",alpha," --b=0.5 --n_burnin=500 --seed=",as.numeric(sub('ENSG','',gene_id)))  
        print(cmd2)
        system(cmd2,ignore.stdout=T,ignore.stderr=T,wait=T)
        
        #load weights generated by prscs
        weights <- fread(paste0(main_dir, "/weights/", gene_id,"_pst_eff_a",alpha,"_b0.5_phiauto_chr",chr,".txt"))   
        
        if (nrow(weights) >0 ){
          
          #generate weight df
          weights$gene <- gene_id
          weights <- weights[,c(7,2,1,3,4,5,6)]
          colnames(weights) <- c("GENE","RSID","CHR","POS","A1","A2","Weight")
          fwrite(weights, row.names = F, col.names = T, sep = "\t", quote = F, file = paste0(main_dir, "/weights/", gene_id,"_weight.txt"))
         
          #generate cov df
          cov_df<-as.data.frame(matrix(data=NA,ncol=4,nrow=1)) 
          colnames(cov_df)<-c('GENE','RSID1','RSID2','VALUE')
          
          #only keep snps in the model for the dosage df
          included_snps_index = sapply(weights$RSID, function(x) which(colnames(dosage) == x))
          counted_allele = counted_allele[included_snps_index]
          dosage <- dosage[,included_snps_index]
          
          if(nrow(weights)>1){
            
            #flipping allele 0 <-> 2 if the counted allele in ref and the effect allele in eQTL doesn't match
            for(f_i in 1:ncol(dosage)){
              if(counted_allele[f_i] != weights[f_i,'A1']){
                index_0 = which(dosage[,f_i] == 0)
                index_2 = which(dosage[,f_i] == 2)
                dosage[index_0,f_i] = 2
                dosage[index_2,f_i] = 0
              }
            }
            
            #calculate covariance
            snp_list<-colnames(dosage)
            o_i=1
            for (k in 1:length(snp_list)){
              for (j in k:length(snp_list)){
                cov_df[o_i,2]<-snp_list[k]
                cov_df[o_i,3]<-snp_list[j]
                cov_df[o_i,4]<-round(cov(dosage[,which(colnames(dosage)==snp_list[k])],dosage[,which(colnames(dosage)==snp_list[j])]),3)
                o_i=o_i+1
              }
            }
            cov_df[,1] <- gene_id
          }else{
            cov_df[,'RSID1'] = cov_df[,'RSID2']=weights[1,'RSID']
            cov_df[,'VALUE'] = cov(dosage, dosage)
            cov_df[,"GENE"] = gene_id
          }
          
          #write covariance
          fwrite(cov_df, sep = "\t", row.names = F, col.names = T, quote = F, file = paste0(main_dir, "/cov/",gene_id,"_covariance.txt"))
          
          #clean up
          cmd3 = paste0('rm ',main_dir,'/gene/',gene_id,'*; rm ',main_dir, "/weights/", gene_id,"_pst_eff_a",alpha,"_b0.5_phiauto_chr",chr,".txt; rm ",main_dir,"/tmp/",gene_id,"_rs_id.txt")
          system(cmd3, wait = T)
          
        }else {
          print("INFO Fail to calculate cov")
          
        }
        
      }else{
        print("INFO ----- Fail to get genotype info")
      }
      
    }else{
      print("INFO ----- Fail to get genotype info")
    }
    
  }else {
    print("INFO ----- Fail to calculate weight")
  }
}


#skip gene already have a model
gene_list_accomplished = dir(paste0(main_dir, "/cov/"), pattern = "*_covariance.txt$",recursive = T)
gene_list_accomplished = sapply(gene_list_accomplished, function(x) sub('_covariance.txt','',x))
skip_index = which(gene_list %in% as.character(gene_list_accomplished))
if(length(skip_index)>0){
  gene_list = gene_list[-skip_index]
}

#start and end id
n = args
i_start <- (n-1)*500+1
if (n < 29){
  i_end <- i_start+499
}else{
  i_end <- 14205
}
print(i_start)
#final gene list need to run
gene_list = gene_list[i_start:i_end]

#run
if(length(unique(gene_list))>2){
  if(parallel){
    registerDoParallel(cores = detectCores())
    #registerDoParallel(cores = max(detectCores()-1, 1)) for vgi01
    foreach(i = 1:length(gene_list)) %dopar% {
      each_gene(gene_list[i])
      
      stopImplicitCluster()
    }
  }else{
    for (i in 1:length(gene_list)){
      print(paste0(i,'/',gene_list[i]))
      each_gene(gene_list[i])
    }
  }
}else{
  print('no more genes')
}



